nottherealalanturing/whattodo
# user authentication

## Description

This pull request adds user authentication to the project using Mongoose, Passport, and JSON Web Tokens (JWT). It includes the following changes:

-     A User model with email and password fields, and pre-save and compare password methods that use bcrypt to hash and compare passwords.

-     A register route that allows users to register with an email and password, and generates a JWT token for the new user.

-     A login route that allows users to log in with an email and password, and generates a JWT token for the authenticated user.

-     A JWT authentication strategy that uses the passport-jwt library to authenticate requests with a JWT token.

-     Error handling for authentication errors.


## Motivation and Context

User authentication is an important feature for many applications, and this pull request adds the necessary functionality to authenticate users with JWT tokens.

## How to Test

To test this pull request, you can run the server and send requests to the /register and /login routes using a tool like curl or Postman. You can use the following example requests:

`POST api/v1/auth/signup

{
  "email": "user@example.com",
  "password": "password123"
}`

`POST api/v1/auth/signin

{
  "email": "user@example.com",
  "password": "password123"
}
`

The /signup route should return a JWT token for the new user, and the /signin route should return a JWT token for the authenticated user. You can use the JWT token to authenticate requests to protected routes by including it in the Authorization header of the request:

`GET /api/v1/auth/me
Authorization: Bearer <jwt-token>
`

closes #7
